File: src/main/scala/com/dyonovan/neotech/registries/SolidifierRecipeHandler.scala
package com.dyonovan.neotech.registries

import java.util

import com.dyonovan.neotech.NeoTech
import com.dyonovan.neotech.managers.MetalManager
import com.google.gson.reflect.TypeToken
import com.teambr.bookshelf.helper.LogHelper
import net.minecraft.command.{ICommandSender, CommandBase}
import net.minecraft.init.{Blocks, Items}
import net.minecraft.item.ItemStack
import net.minecraft.server.MinecraftServer
import net.minecraft.util.text.TextComponentString
import net.minecraft.util.text.translation.I18n
import net.minecraftforge.fluids.{FluidRegistry, FluidStack}
import net.minecraftforge.oredict.OreDictionary

/**
  * This file was created for NeoTech
  *
  * NeoTech is licensed under the
  * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
  * http://creativecommons.org/licenses/by-nc-sa/4.0/
  *
  * @author Paul Davis <pauljoda>
  * @since 2/18/2016
  */
class SolidifierRecipeHandler extends AbstractRecipeHandler[SolidifierRecipe, FluidStack, ItemStack] {

    /**
      * Used to get the base name of the files
      *
      * @return
      */
    override def getBaseName: String = "solidifier"

    /**
      * This is the current version of the registry, if you update this it will cause the registry to be redone
      *
      * @return
      */
    override def getVersion: Int = 4

    /**
      * Used to get the default folder location
      *
      * @return
      */
    override def getBaseFolderLocation: String = NeoTech.configFolderLocation

    /**
      * Used to get what type token to read from file (Generics don't handle well)
      *
      * @return
      */
    override def getTypeToken: TypeToken[util.ArrayList[SolidifierRecipe]] =
        new TypeToken[util.ArrayList[SolidifierRecipe]]() {}

    /**
      * Used to generate the default values
      */
    override def generateDefaultRecipes(): Unit = {
        LogHelper.info("Json not found. Creating Dynamic Solidifier Recipe List...")

        // Metals
        val iterator = MetalManager.metalRegistry.keySet().iterator()
        while(iterator.hasNext) {
            val metal = MetalManager.metalRegistry.get(iterator.next())
            if (FluidRegistry.isFluidRegistered(metal.oreDict)) {

                //Block
                if(metal.block.isDefined)
                    addSolidifierRecipe(new FluidStack(FluidRegistry.getFluid(metal.oreDict), MetalManager.BLOCK_MB), null, metal.block.get.getName)

                //Ingot
                if(metal.ingot.isDefined)
                    addSolidifierRecipe(new FluidStack(FluidRegistry.getFluid(metal.oreDict), MetalManager.INGOT_MB), null, metal.ingot.get.getName)

                //Nugget
                if(metal.nugget.isDefined)
                    addSolidifierRecipe(new FluidStack(FluidRegistry.getFluid(metal.oreDict), MetalManager.NUGGET_MB), null, metal.nugget.get.getName)
            }
        }

        // Extend Vanilla

        // Iron
        addSolidifierRecipe(new FluidStack(FluidRegistry.getFluid("iron"), MetalManager.BLOCK_MB),
            new ItemStack(Blocks.iron_block), "blockIron")
        addSolidifierRecipe(new FluidStack(FluidRegistry.getFluid("iron"), MetalManager.INGOT_MB),
            new ItemStack(Items.iron_ingot), "ingotIron")

        // Gold
        addSolidifierRecipe(new FluidStack(FluidRegistry.getFluid("gold"), MetalManager.BLOCK_MB),
            new ItemStack(Blocks.gold_block), "blockGold")
        addSolidifierRecipe(new FluidStack(FluidRegistry.getFluid("gold"), MetalManager.INGOT_MB),
            new ItemStack(Items.gold_ingot), "ingotGold")
        addSolidifierRecipe(new FluidStack(FluidRegistry.getFluid("gold"), MetalManager.NUGGET_MB),
            new ItemStack(Items.gold_nugget), "nuggetGold")

        // Carbon
        addSolidifierRecipe(new FluidStack(FluidRegistry.getFluid("carbon"), MetalManager.INGOT_MB),
            new ItemStack(Items.coal), "")
        addSolidifierRecipe(new FluidStack(FluidRegistry.getFluid("carbon"), MetalManager.BLOCK_MB),
            new ItemStack(Blocks.coal_block), "")

        // Obsidian
        addSolidifierRecipe(new FluidStack(FluidRegistry.getFluid("obsidian"), 36),
            new ItemStack(Blocks.obsidian), "")

        saveToFile()
        LogHelper.info("Finished adding " + recipes.size + " Solidifier Recipes")
    }

    /**
      * Adds the recipe
      *
      * @param output If you set null for the itemstack, it will attempt to create one from ore dict
      * @param fluidStack FluidStack
      */
    def addSolidifierRecipe(fluidStack: FluidStack, output : ItemStack, ore : String) : Unit = {
        var stack : ItemStack = output
        if(output == null && !ore.isEmpty) {
            val stackList = OreDictionary.getOres(ore)
            if(!stackList.isEmpty) {
                stack = stackList.get(0)
            } else {
                LogHelper.severe("Could not add ore dict solidifier recipe for " + ore + " as it does not exist in the OreDictionary")
                return
            }
        }
        val recipe = new SolidifierRecipe(getFluidString(fluidStack), ore, getItemStackString(stack))
        addRecipe(recipe)
    }

    /**
      * Get the command to add values to the registry
      *
      * @return A new command
      */
    override def getCommand: CommandBase = {
        new CommandBase {
            override def getCommandName: String = "addSolidifierRecipe"

            override def getRequiredPermissionLevel : Int = 3

            override def getCommandUsage(sender: ICommandSender): String = "commands.addSolidifierRecipe.usage"

            override def execute(server: MinecraftServer, sender: ICommandSender, args: Array[String]): Unit = {
                if(args.length < 2)
                    sender.addChatMessage(new TextComponentString(I18n.translateToLocal("commands.addSolidifierRecipe.usage")))
                else {
                    var outputStack : String = null
                    if(args(1).split(":").nonEmpty) {
                        outputStack = args(1)
                        if(getItemStackFromString(outputStack) != null && getFluidFromString(args(0)) != null) {
                            addRecipe(new SolidifierRecipe(args(0), "", outputStack))
                            sender.addChatMessage(new TextComponentString(args(0) + " -> " + outputStack + " Added Successfully"))
                            saveToFile()
                        } else
                            sender.addChatMessage(new TextComponentString(args(0) + " -> " + outputStack + " Failed to be added"))
                    } else {
                        if(!OreDictionary.getOres(args(1)).isEmpty && getFluidFromString(args(0)) != null) {
                            addRecipe(new SolidifierRecipe(args(0), args(1), ""))
                            sender.addChatMessage(new TextComponentString(args(0) + " -> " + args(1) + " Added Successfully"))
                            saveToFile()
                        } else
                            sender.addChatMessage(new TextComponentString(args(0) + " -> " + outputStack + " Failed to be added"))
                    }
                }
            }
        }
    }
}


/**
  * Helper class for holding recipes
  *
  */
class SolidifierRecipe(val input : String, val ore : String, val output : String)
        extends AbstractRecipe[FluidStack, ItemStack] {
    /**
      * Used to get the output of this recipe
      *
      * @param fluidIn The input object
      * @return The output object
      */
    override def getOutput(fluidIn: FluidStack): Option[ItemStack] = {
        if(fluidIn == null)
            return None
        if(getFluidFromString(input).isFluidEqual(fluidIn) && getFluidFromString(input).amount == fluidIn.amount) {
            if(getItemStackFromString(output) != null)
                return Option(getItemStackFromString(output))
            else {
                if(ore != null && !OreDictionary.getOres(ore).isEmpty) {
                    return Option(OreDictionary.getOres(ore).get(0))
                }
            }
        }
        None
    }

    /**
      * Is the input valid for an output
      *
      * @param fluidIn The input object
      * @return True if there is an output
      */
    override def isValidInput(fluidIn: FluidStack): Boolean = {
        if(fluidIn == null || fluidIn.getFluid == null)
            return false
        getFluidFromString(input).getFluid == fluidIn.getFluid
    }
}

File: src/main/java/com/teambrmodding/neotech/registries/SolidifierRecipeHandler.java
package com.teambrmodding.neotech.registries;

import com.google.gson.reflect.TypeToken;
import com.teambr.bookshelf.helper.LogHelper;
import com.teambrmodding.neotech.Neotech;
import com.teambrmodding.neotech.collections.SolidifierMode;
import com.teambrmodding.neotech.managers.MetalManager;
import com.teambrmodding.neotech.registries.recipes.SolidifierRecipe;
import net.minecraft.command.CommandBase;
import net.minecraft.command.CommandException;
import net.minecraft.command.ICommandSender;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.init.Blocks;
import net.minecraft.init.Items;
import net.minecraft.item.ItemStack;
import net.minecraft.server.MinecraftServer;
import net.minecraft.util.text.TextComponentString;
import net.minecraft.util.text.translation.I18n;
import net.minecraftforge.fluids.FluidRegistry;
import net.minecraftforge.fluids.FluidStack;
import net.minecraftforge.fluids.capability.CapabilityFluidHandler;
import net.minecraftforge.fluids.capability.IFluidHandler;
import net.minecraftforge.fluids.capability.IFluidTankProperties;
import org.apache.commons.lang3.tuple.Pair;

import java.util.ArrayList;

/**
 * This file was created for NeoTech
 *
 * NeoTech is licensed under the
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
 * http://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * @author Paul Davis - pauljoda
 * @since 2/15/2017
 */
public class SolidifierRecipeHandler extends
        AbstractRecipeHandler<SolidifierRecipe, Pair<SolidifierMode, FluidStack>, ItemStack> {

    /**
     * Used to get the base name of the files
     *
     * @return The base name for the files
     */
    @Override
    public String getBaseName() {
        return "solidifier";
    }

    /**
     * This is the current version of the registry, if you update this it will cause the registry to be redone
     *
     * @return Current version number
     */
    @Override
    public int getVersion() {
        return 0;
    }

    /**
     * Used to get the default folder location
     *
     * @return The folder location
     */
    @Override
    public String getBaseFolderLocation() {
        return Neotech.configFolderLocation;
    }

    /**
     * Used to get what type token to read from file (Generics don't handle well)
     *
     * @return A defined type token
     */
    @Override
    public TypeToken<ArrayList<SolidifierRecipe>> getTypeToken() {
        return new TypeToken<ArrayList<SolidifierRecipe>>() {
        };
    }

    /**
     * Get the command to add values to the registry
     *
     * @return A new command
     */
    @Override
    public CommandBase getCommand() {
        return new CommandBase() {
            @Override
            public String getCommandName() {
                return "addSolidifierRecipe";
            }

            @Override
            public int getRequiredPermissionLevel() {
                return 3;
            }

            @Override
            public String getCommandUsage(ICommandSender sender) {
                return "commands.addSolidifierRecipe.usage";
            }

            @Override
            public void execute(MinecraftServer server, ICommandSender sender, String[] args) throws CommandException {
                if (args.length != 3 || (args.length == 2 && !args[0].equalsIgnoreCase("hands")))
                    sender.addChatMessage(new TextComponentString(I18n.translateToLocal(getCommandUsage(sender))));
                else if (args.length == 2 && args[0].equalsIgnoreCase("hands")) { // Allow user to hold recipe

                    String mode = args[1];
                    SolidifierMode requiredMode = null;
                    if (mode.equalsIgnoreCase("BLOCK"))
                        requiredMode = SolidifierMode.BLOCK_MODE;
                    else if (mode.equalsIgnoreCase("INGOT"))
                        requiredMode = SolidifierMode.INGOT_MODE;
                    else if (mode.equalsIgnoreCase("NUGGET"))
                        requiredMode = SolidifierMode.NUGGET_MODE;
                    else {
                        // Conditions for hands usage not met
                        sender.addChatMessage(new TextComponentString(I18n.translateToLocal(getCommandUsage(sender))));
                        return;
                    }

                    // Must be a player using the command as we need their hands
                    if (sender.getCommandSenderEntity() instanceof EntityPlayer) {
                        EntityPlayer player = (EntityPlayer) sender.getCommandSenderEntity(); // Get the player

                        // Make sure both hands have a stack
                        if (player.getHeldItemMainhand() != null && player.getHeldItemOffhand() != null) {
                            ItemStack mainHandStack = player.getHeldItemMainhand();
                            ItemStack offHandStack = player.getHeldItemOffhand();

                            // Offhand must hold a fluid handler
                            if (offHandStack.hasCapability(CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY, null)) {
                                IFluidHandler fluidHandler = offHandStack.getCapability(CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY, null);

                                // Cycle the tanks (usually one) to find a tank with a fluid
                                for (IFluidTankProperties tankInfo : fluidHandler.getTankProperties()) {
                                    FluidStack stackInTank = tankInfo.getContents();

                                    // Is a valid FluidStack
                                    if (stackInTank != null && stackInTank.getFluid() != null) {
                                        addRecipe(new SolidifierRecipe(requiredMode, getItemStackString(mainHandStack), getFluidStackString(stackInTank)));
                                        sender.addChatMessage(new TextComponentString(getItemStackString(mainHandStack) + " -> " +
                                                getFluidStackString(stackInTank) + " Added Successfully!"));
                                        saveToFile();
                                        return;
                                    }
                                }
                            }
                        }
                    }

                    // Conditions for hands usage not met
                    sender.addChatMessage(new TextComponentString(I18n.translateToLocal(getCommandUsage(sender))));
                } else { // Not hands, has three args
                    String fluidStackInput = args[1];
                    String itemStackOutput = args[2];

                    String mode = args[0];
                    SolidifierMode requiredMode = null;
                    if (mode.equalsIgnoreCase("BLOCK"))
                        requiredMode = SolidifierMode.BLOCK_MODE;
                    else if (mode.equalsIgnoreCase("INGOT"))
                        requiredMode = SolidifierMode.INGOT_MODE;
                    else if (mode.equalsIgnoreCase("NUGGET"))
                        requiredMode = SolidifierMode.NUGGET_MODE;
                    else {
                        // Conditions for hands usage not met
                        sender.addChatMessage(new TextComponentString(I18n.translateToLocal(getCommandUsage(sender))));
                        return;
                    }

                    if (getFluidStackFromString(fluidStackInput) != null && getItemStackFromString(itemStackOutput) != null) {
                        addRecipe(new SolidifierRecipe(requiredMode, fluidStackInput, itemStackOutput));
                        sender.addChatMessage(new TextComponentString(fluidStackInput + " -> " + itemStackOutput + " Added Successfully!"));
                        saveToFile();
                    } else
                        sender.addChatMessage(new TextComponentString(I18n.translateToLocal(getCommandUsage(sender))));
                }
            }
        };
    }

    @Override
    protected void generateDefaultRecipes() {
        LogHelper.logger.info("[Neotech] Generating Solidifier Recipes...");

        // Metals
        for (String key : MetalManager.metalRegistry.keySet()) {
            MetalManager.Metal metal = MetalManager.getMetal(key);
            if (FluidRegistry.isFluidRegistered(metal.getOreDict())) {
                // Block
                if (metal.getSolidBlock() != null)
                    addRecipe(new SolidifierRecipe(SolidifierMode.BLOCK_MODE, metal.getOreDict() + ":" + MetalManager.BLOCK_MB,
                            getItemStackString(new ItemStack(metal.getSolidBlock(), 1))));

                // Ingot
                if (metal.getIngot() != null)
                    addRecipe(new SolidifierRecipe(SolidifierMode.INGOT_MODE, metal.getOreDict() + ":" + MetalManager.INGOT_MB,
                            getItemStackString(new ItemStack(metal.getIngot(), 1))));

                // Nugget
                if (metal.getNugget() != null)
                    addRecipe(new SolidifierRecipe(SolidifierMode.NUGGET_MODE, metal.getOreDict() + ":" + MetalManager.NUGGET_MB,
                            getItemStackString(new ItemStack(metal.getNugget(), 1))));
            }
        }

        // Vanilla Stuff

        //Iron
        addRecipe(new SolidifierRecipe(SolidifierMode.BLOCK_MODE, "iron:" + MetalManager.BLOCK_MB, "blockIron"));
        addRecipe(new SolidifierRecipe(SolidifierMode.INGOT_MODE, "iron:" + MetalManager.INGOT_MB, "ingotIron"));

        // Gold
        addRecipe(new SolidifierRecipe(SolidifierMode.BLOCK_MODE, "gold:" + MetalManager.BLOCK_MB, "blockGold"));
        addRecipe(new SolidifierRecipe(SolidifierMode.INGOT_MODE, "gold:" + MetalManager.INGOT_MB, "ingotGold"));
        addRecipe(new SolidifierRecipe(SolidifierMode.NUGGET_MODE, "gold:" + MetalManager.NUGGET_MB, "nuggetGold"));

        // Carbon
        addRecipe(new SolidifierRecipe(SolidifierMode.BLOCK_MODE, "carbon:" + MetalManager.BLOCK_MB,
                getItemStackString(new ItemStack(Blocks.COAL_BLOCK, 1))));
        addRecipe(new SolidifierRecipe(SolidifierMode.INGOT_MODE, "carbon:" + MetalManager.INGOT_MB,
                getItemStackString(new ItemStack(Items.COAL, 1))));

        // Obsidian
        addRecipe(new SolidifierRecipe(SolidifierMode.BLOCK_MODE, "obsidian:288",
                getItemStackString(new ItemStack(Blocks.OBSIDIAN, 1))));

        // Blaze
        addRecipe(new SolidifierRecipe(SolidifierMode.INGOT_MODE, "blaze:" + MetalManager.NUGGET_MB,
                getItemStackString(new ItemStack(Items.BLAZE_ROD))));

        // Chorus
        addRecipe(new SolidifierRecipe(SolidifierMode.BLOCK_MODE, "chorus:" + MetalManager.NUGGET_MB,
                getItemStackString(new ItemStack(Items.CHORUS_FRUIT_POPPED))));

        // Wither
        addRecipe(new SolidifierRecipe(SolidifierMode.NUGGET_MODE, "wither:" + MetalManager.NUGGET_MB,
                getItemStackString(new ItemStack(Items.NETHER_STAR))));

        saveToFile();
    }
}